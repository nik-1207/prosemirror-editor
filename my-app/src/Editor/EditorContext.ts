import React from "react";

export const EditorContext: any = React.createContext(null);
