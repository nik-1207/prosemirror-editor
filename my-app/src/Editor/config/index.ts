import { keymap } from "prosemirror-keymap";
import { history } from "prosemirror-history";
import { baseKeymap } from "prosemirror-commands";
import { Plugin } from "prosemirror-state";
import { menuBar } from "prosemirror-menu";

import { buildMenuItems } from "./menu";
import { buildKeymap } from "./keymap";
import { buildInputRules } from "./inputrules";

export function setup(options: any) {
  let plugins = [
    buildInputRules(options.schema),
    keymap(buildKeymap(options.schema, options.mapKeys)),
    keymap(baseKeymap),
  ];
  if (options.menuBar !== false)
    plugins.push(
      //@ts-ignore
      menuBar({
        content: buildMenuItems(options.schema).fullMenu,
      })
    );
  if (options.history !== false) plugins.push(history());

  return plugins.concat(
    new Plugin({
      props: {
        attributes: { class: "ProseMirror-example-setup-style" },
      },
    })
  );
}
