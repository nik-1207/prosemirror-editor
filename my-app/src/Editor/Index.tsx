import React from "react";

import { DOMParser } from "prosemirror-model";
import { EditorState } from "prosemirror-state";

import { EditorContext } from "./EditorContext";

import "./style.css";

interface ProsemirrorProps {
  uniqueId: string | number;
}

const Editor: React.FC<ProsemirrorProps> = ({ uniqueId }) => {
  const { editor, view, schema } = React.useContext(EditorContext);
  React.useEffect(() => {
    const wrapper = document.getElementById(`wrapper${uniqueId}`)!;
    wrapper.appendChild(editor);
    return () => {
      view.updateState(
        EditorState.create({
          doc: DOMParser.fromSchema(schema).parse(
            document.createElement("div")
          ),
          plugins: view.state.plugins,
        })
      );
    };
  }, [uniqueId]); // eslint-disable-line

  return <div id={`wrapper${uniqueId}`} />;
};

export default Editor;
