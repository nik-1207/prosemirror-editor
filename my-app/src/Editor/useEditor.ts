import { DOMParser, Node as prosemirrorNode, Schema } from "prosemirror-model";
import { EditorState } from "prosemirror-state";
import { EditorView } from "prosemirror-view";
import { schema as baseSchema } from "./config/prosemirror-schema-basic";
import { addListNodes } from "./config/prosemirror-schema-list";

import { setup } from "./config";

const useEditor = (checklist?: string) => {
  const editor = document.createElement("div");
  const content = document.createElement("div");
  content.id = "content";
  editor.id = "editor";

  const schema = new Schema({
    nodes: addListNodes(baseSchema.spec.nodes, "note block*", "block"),
    marks: baseSchema.spec.marks,
  });

  const state = EditorState.create({
    doc: DOMParser.fromSchema(schema).parse(content as Node),
    plugins: setup({ schema }),
  });

  let view = new EditorView(editor as Node, {
    state,
    dispatchTransaction: (transaction) => {
      let newState = view.state.apply(transaction);
      view.updateState(newState);
    },
  });
  if (checklist) {
    const jsonData = JSON.parse(checklist);
    const node = prosemirrorNode.fromJSON(schema, jsonData);
    view.dispatch(view.state.tr.replaceSelectionWith(node));
  }
  return {
    editor,
    view,
    state,
    schema,
  };
};
export { useEditor };
