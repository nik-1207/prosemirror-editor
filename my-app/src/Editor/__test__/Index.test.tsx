import { render } from "@testing-library/react";

import Editor from "../Index";
import { EditorContext } from "../EditorContext";
import { useEditor } from "../useEditor";

describe("Editor", () => {
  it("should match snapshot", () => {
    const EditorWrapper: React.FC = () => {
      const { editor, view, schema } = useEditor();
      return (
        <EditorContext.Provider value={{ editor, view, schema }}>
          <Editor uniqueId="test-checklist-uid" />
        </EditorContext.Provider>
      );
    };

    const { container } = render(<EditorWrapper />);

    expect(container).toMatchSnapshot();
  });
});
