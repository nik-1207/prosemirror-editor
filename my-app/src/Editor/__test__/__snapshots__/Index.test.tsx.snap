// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should match snapshot 1`] = `
<div>
  <div
    id="wrappertest-checklist-uid"
  >
    <div
      id="editor"
    >
      <div
        class="ProseMirror-menubar-wrapper"
      >
        <div
          class="ProseMirror-menubar"
        >
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Toggle strong style"
            >
              <svg
                style="width: 0.7861328125em;"
              >
                <use
                  href="http://localhost/#pm-icon-4eb53ba9"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Toggle emphasis"
            >
              <svg
                style="width: 0.5712890625em;"
              >
                <use
                  href="http://localhost/#pm-icon-25612f7a"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Toggle code font"
            >
              <svg
                style="width: 0.875em;"
              >
                <use
                  href="http://localhost/#pm-icon--1d0983fe"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon ProseMirror-menu-disabled"
              title="Add or remove link"
            >
              <svg
                style="width: 0.9287109375em;"
              >
                <use
                  href="http://localhost/#pm-icon--78ecc05e"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuseparator"
          />
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-menu-dropdown-wrap"
            >
              <div
                class="ProseMirror-menu-dropdown "
              >
                Insert
              </div>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-menu-dropdown-wrap"
            >
              <div
                class="ProseMirror-menu-dropdown "
              >
                Type...
              </div>
            </div>
          </span>
          <span
            class="ProseMirror-menuseparator"
          />
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon ProseMirror-menu-disabled"
              title="Undo last change"
            >
              <svg
                style="width: 1em;"
              >
                <use
                  href="http://localhost/#pm-icon-5a60796"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon ProseMirror-menu-disabled"
              title="Redo last undone change"
            >
              <svg
                style="width: 1em;"
              >
                <use
                  href="http://localhost/#pm-icon--6ff882d4"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuseparator"
          />
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Wrap in bullet list"
            >
              <svg
                style="width: 0.8571428571428571em;"
              >
                <use
                  href="http://localhost/#pm-icon--6141f81"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Wrap in ordered list"
            >
              <svg
                style="width: 0.8571428571428571em;"
              >
                <use
                  href="http://localhost/#pm-icon-28a6efee"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Wrap in block quote"
            >
              <svg
                style="width: 0.7142857142857143em;"
              >
                <use
                  href="http://localhost/#pm-icon--2d1566b3"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
            style="display: none;"
          >
            <div
              class="ProseMirror-icon"
              style="display: none;"
              title="Join with above block"
            >
              <svg
                style="width: 0.8888888888888888em;"
              >
                <use
                  href="http://localhost/#pm-icon--3cb9c8"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
            style="display: none;"
          >
            <div
              class="ProseMirror-icon"
              style="display: none;"
              title="Lift out of enclosing block"
            >
              <svg
                style="width: 1em;"
              >
                <use
                  href="http://localhost/#pm-icon--2eec1b76"
                />
              </svg>
            </div>
          </span>
          <span
            class="ProseMirror-menuitem"
          >
            <div
              class="ProseMirror-icon"
              title="Select parent node"
            >
              <span
                style="font-weight: bold;"
              >
                â¬š
              </span>
            </div>
          </span>
        </div>
        <div
          class="ProseMirror ProseMirror-example-setup-style"
          contenteditable="true"
          translate="no"
        >
          <p>
            <br
              class="ProseMirror-trailingBreak"
            />
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
